torch/quantization/fx/quantize.py

```diff
@@ -407,6 +414,696 @@ def load_arg(a):
+        # print('\nnode_name_to_target_dtype', node_name_to_target_dtype)
+        # print('processing node', node.format_node(), '\n')
+
+        # copy the current node as is
+        new_node = observed_graph.node_copy(node, load_arg)
+        env[node.name] = new_node
+
+        # check for matches
+        root_node, matched_nodes, pattern, qhandler, qconfig = matches.get(
+            node.name, (None, None, None, None, None))
+        # print('match results', root_node, matched_nodes, pattern, qhandler, qconfig)
+
+        #
+        # After this point, the current node and all of its arguments
+        # have a dtype assigned. Now, we insert observers for inputs
+        # of this node (if needed for this node), and the output of this node
+        # (if needed for this node).
+        #
+
+        if new_node.op == 'placeholder':
+            # if a graph input is in fp32, it does not need observation
+            # if a graph input is in int8, we assume the observation happens
+            #   outside of the graph, and no additional observation is needed
+            pass
+
+        elif new_node.op in ('call_module', 'call_method', 'call_function', 'output'):
+            modules = dict(model.named_modules(remove_duplicate=False))
+            this_node_dtype = node_name_to_target_dtype[new_node.name]
+            output_not_a_tensor = this_node_dtype is None
+            # TODO(future PR): consider stopping matching getitem
+            is_getitem = new_node.op == 'call_function' and \
+                new_node.target == operator.getitem
+
+            skip_inserting_observers = (
+                (qconfig is None) or
+                output_not_a_tensor or
+                is_getitem
+            ) and (not new_node.op == 'output')
+
+            if not skip_inserting_observers:
+                if new_node.op != 'output':
+                    # this modifies new_node inplace
+                    maybe_insert_input_observers_for_node_rewrite(
+                        new_node, qconfig, load_arg, model, modules, observed_graph,
+                        node_name_to_target_dtype,
+                        qhandler, prepare_custom_config_dict)
+
+                    is_last_node_of_pattern = root_node is node
+                    is_like_copy_node = \
+                        (qhandler is not None and (
+                            isinstance(qhandler, CopyNodeQuantizeHandler)
+                        ))
+                    if is_last_node_of_pattern and (not is_like_copy_node):
+                        # this returns the new observer node if it was needed
+                        maybe_output_obs_node = maybe_insert_output_observer_for_node_rewrite(
+                            new_node, model, modules, observed_graph, matches,
+                            node_name_to_target_dtype, pattern, qhandler)
+                        if maybe_output_obs_node is not None:
+                            env[new_node.name] = maybe_output_obs_node
```

Comment:
looks like we updated the value for new_node.name here, is supriya's comment address in the PR? https://github.com/pytorch/pytorch/pull/57519#issuecomment-840094762

Reply:
yes, this should work without changes.  The key distinction is this only applies to node output, and a node can only have a single output dtype of the first argument after convert.  

for example,

```
x0_fp32 -> obs0 -> int8_op
\
fp32_op
```

in the graph above, x0_fp32's target dtype is fp32, so no observer will be inserted to its output.  `obs0` will be inserted when the `int8_op` node is processed.  This is a key difference of the proposal in this PR (each node inserts input and output observers it needs) versus the previous behavior (observers are always inserted to the output of current node, even if they are only needed for a future node)

Reply:
or, to explain it in a different way:
* a node can only have one output type after convert
* the "multiple observers of a single node" use case comes from nodes of different target dtypes having a shared node as an input
* the observers for inputs of a node are not handled in this code path (and they are inserted without modifying `env`)